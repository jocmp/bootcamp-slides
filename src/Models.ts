export interface SlideshowModel {
    id: number;
    title: string;
    slides: SlideModel[];
}

export interface SlideModel {
    title: string;
    slide_type: string;
    content?: string[];
}

export interface AppState {
    slideshow: SlideshowModel;
}