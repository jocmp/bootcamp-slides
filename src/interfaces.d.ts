declare module 'enzyme-to-json'