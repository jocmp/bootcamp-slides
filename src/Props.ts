import { SlideModel, SlideshowModel } from './Models'

export interface SlideshowOwnProps {
    history: any;
    match: { params: any };
}

export interface SlideshowProps extends SlideshowStateProps, SlideshowDispatchProps, SlideshowOwnProps { };

export interface SlideshowStateProps {
    slideshow: SlideshowModel;
    viewedIndices: number[];
}

export interface SlideshowDispatchProps {
    handleNextSlide: (id: string, index: string, length: number, history: any) => any;
    handlePreviousSlide: (id: string, index: string, length: number, history: any) => any;
    fetchSlideshow: (id: number, index: number) => void;
    viewSlide: (id: number) => any;
}

export interface OverviewProps {
    match: { params: any };
    slideshow: SlideshowModel;
    viewedIndices: number[];
    viewSlide: (id: number) => any;
}

export interface OverviewItemProps {
    slideshow_id: number;
    index: number;
    slide: SlideModel;
    viewed: boolean;
    isCurrent: boolean;
    viewSlide: (id: number) => any;
}

export interface SlideProps {
    slide: SlideModel;
}
