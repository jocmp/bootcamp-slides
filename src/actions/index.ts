import { SlideshowModel } from '../Models'
import { History } from 'history';

export const nextSlide = (id: string, index: string, totalLength: number, history: History) => {
    const currentIndex = parseInt(index);
    const nextIndex: number = currentIndex < totalLength - 1 ? currentIndex + 1 : currentIndex;
    return handleSlide(id, nextIndex, history);
}

export const previousSlide = (id: string, index: string, totalLength: number, history: History) => {
    const currentIndex = parseInt(index);
    const previousIndex: number = currentIndex > 0 ? currentIndex - 1 : currentIndex;
    return handleSlide(id, previousIndex, history);
}

const handleSlide = (id: string, index: number, history: History) => {
    history.push(`/slideshows/${id}/slides/${index}`);
    return viewSlide(index);
}

export const viewSlide = (index: number) => {
    return {
        type: 'VIEW_SLIDE',
        index
    };
}

const apiData = (index: number): SlideshowModel => {
    const shows: SlideshowModel[] = [
        {
            id: 1,
            title: "Exceptional Presentation",
            slides: [
                {
                    title: "Title here",
                    slide_type: "title"
                },
                {
                    title: "Sensational Informational",
                    slide_type: "simple",
                    content: ["Inspirational insatiable"]
                }
            ]
        },
        {
            id: 2,
            title: "Touch Fuzzy, Get Dizzy",
            slides: [
                {
                    title: "Yoshi's Island",
                    slide_type: "title",
                },
                {
                    title: "Yoshi's Island",
                    slide_type: "simple",
                    content: ['Imperically the best Super Mario game... \ntest']
                },
                {
                    title: "Why buy a SNES?",
                    slide_type: "simple",
                    content: ["Yoshi's Island. No question."]
                }
            ]
        }
    ]
    return shows[index - 1];
};

export const loadSlideshow = (slideshow: SlideshowModel, currentIndex: number) => {
    return {
        type: 'LOAD_SLIDESHOW',
        slideshow,
        currentIndex
    }
};

export const fetchSlideshow = (id: number, index: number = 0) => (dispatch: any, getState: any) => {
    dispatch(loadSlideshow(apiData(id), index));
};