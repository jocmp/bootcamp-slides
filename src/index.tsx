import * as React from 'react';
import * as ReactDOM from 'react-dom';
import App from './components/App';
import './index.css';
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

export const slideshow = (state: any, action: any) => {
  switch(action.type) {
    case 'LOAD_SLIDESHOW':
      return {
        slideshow: action.slideshow
      }
    default:
      return {};
  }
}

export const store = createStore(slideshow, applyMiddleware(thunk));
store.subscribe(() => console.log(store.getState()))

ReactDOM.render(
  <App store={store} />,
  document.getElementById('root') as HTMLElement
);
