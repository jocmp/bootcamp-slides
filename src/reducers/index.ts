import { combineReducers } from 'redux'
import { AppState } from '../Models'

const viewedIndices = (state: number[] = [], action: any) => {
    switch (action.type) {
        case 'VIEW_SLIDE':
            return Array.from(new Set([...state, action.index]));
        default:
            return state;
    }
};

const initialSlideshow: any = { title: "", id: -1, slides: [] };

const slideshow = (state: AppState, action: any) => {
    switch (action.type) {
        case 'LOAD_SLIDESHOW':
            return action.slideshow;
        case 'LOAD_SLIDESHOW_ERROR':
            return initialSlideshow;
        case 'VIEW_SLIDE':
        default:
            if (!!state) {
                return state;
            }
            return initialSlideshow;
    }
}

const loading = (state: AppState, action: any) => {
    switch (action.type) {
        case 'FETCH_SLIDESHOW':
            return action.loading;
        default:
            return loading;
    }
}

const error = (state: AppState, action: any) => {
    switch (action.type) {
        case 'LOAD_SLIDESHOW_ERROR':
            return action.error;
        case 'CLEAR_ERROR':
            return '';
        default:
            return '';
    }
}


const rootReducer = combineReducers({
    slideshow,
    viewedIndices,
    loading,
    error
});

export default rootReducer;