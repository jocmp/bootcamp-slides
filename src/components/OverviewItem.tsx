import * as React from 'react'
import { OverviewItemProps } from '../Props'
import '../stylesheets/Overview.scss'
import { Link } from 'react-router-dom'

const OverviewItem = (props: OverviewItemProps) => {
    return (
        <li className={props.viewed ? "overview-item-viewed" : "overview-item"}>
            <Link onClick={() => props.viewSlide(props.index)} to={`/slideshows/${props.slideshow_id}/slides/${props.index}`}>
                {props.slide.title}
            </Link>
        </li>
    );
}

export default OverviewItem;