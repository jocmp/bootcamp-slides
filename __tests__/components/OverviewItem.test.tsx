import * as React from 'react'
import OverviewItem from '../../src/components/OverviewItem'
import { SlideModel } from '../../src/Models'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('OverviewItem', () => {

    const slide: SlideModel = {
        title: "Exciting Slide",
        slide_type: "title",
    }; 

    it('renders', () => {
        const wrapper = shallow(
            <OverviewItem viewSlide={jest.fn()} viewed={false} slide={slide} slideshow_id={1} index={0} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    });
});