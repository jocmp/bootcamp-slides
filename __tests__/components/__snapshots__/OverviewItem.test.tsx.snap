// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewItem renders 1`] = `
<li
  className="overview-item"
>
  <Link
    onClick={[Function]}
    replace={false}
    to="/slideshows/1/slides/0"
  >
    Exciting Slide
  </Link>
</li>
`;
